crates/admin/src/cluster_controller/nodeset_selection.rs

```diff
@@ -0,0 +1,541 @@
+                .keys()
+                .copied()
+                .filter(|node_id| {
+                    nodes_config
+                        .get_log_server_storage_state(node_id)
+                        .can_write_to()
+                })
+                .collect(),
+        )
+    }
+
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+}
+
+/// Nodeset selector for picking a set of storage nodes for a replicated loglet out of a
+/// broader pool of available nodes.
+///
+/// This selector can be reused once constructed to make multiple decision in a single scheduling iteration,
+/// if the node configuration and the replication settings are not changing.
+#[cfg(feature = "replicated-loglet")]
+#[derive(Clone)]
+pub struct NodeSetSelector<'a> {
+    nodes_config: &'a NodesConfiguration,
+    cluster_state: &'a ObservedClusterState,
+}
+
+impl<'a> NodeSetSelector<'a> {
+    pub fn new(
+        nodes_config: &'a NodesConfiguration,
+        cluster_state: &'a ObservedClusterState,
+    ) -> NodeSetSelector<'a> {
+        Self {
+            nodes_config,
+            cluster_state,
+        }
+    }
+
+    /// Picks a set of storage nodes for a replicated loglet out of the available pool. Only alive, writable
+    /// storage nodes will be used.
+    pub fn select<R: Rng + ?Sized>(
+        &self,
+        strategy: NodeSetSelectionStrategy,
+        replication_property: &ReplicationProperty,
+        rng: &mut R,
+        preferred_nodes: &NodeSet,
+    ) -> Result<NodeSet, NodeSelectionError> {
+        if replication_property.at_greatest_scope().0 != &LocationScope::Node {
+            // todo: add support for other location scopes
+            unimplemented!("only node-scoped replication is currently supported");
+        }
+
+        // Only consider alive, writable storage nodes.
+        let candidates = WritableNodeSet::from(self.cluster_state, self.nodes_config);
+
+        let min_copies = replication_property.num_copies();
+        if candidates.len() < min_copies.into() {
+            trace!(
+                candidate_nodes_count = ?candidates.len(),
+                ?min_copies,
+                cluster_state = ?self.cluster_state,
+                nodes_config = ?self.nodes_config,
+                "Not enough writeable nodes to meet the minimum replication requirements"
+            );
+            return Err(NodeSelectionError::InsufficientWriteableNodes);
+        }
+
+        // ReplicationFactor(f+1) implies a minimum of 2f+1 nodes. At this point we are only
+        // calculating the nodeset floor size, the actual size will be determined by the specific
+        // strategy in use. This is mainly to prevent math overflow during the size calculation.
+        assert!(
+            min_copies < u8::MAX >> 1,
+            "The replication factor implies a cluster size that exceeds the maximum supported size"
+        );
+        let optimal_fault_tolerant_nodeset_size = (min_copies as usize - 1) * 2 + 1;
```

Comment:
nit: I prefer `usize::from(min_copies)` over `min_copies as usize` because it gives us compiler support whenever we change the type of `min_copies`. With `as` the compiler would also do conversions that are not reversible (e.g. `u32::MAX as u8`).